cpu/atmega328p/periph/uart.c

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright (C) 2014 Freie Universität Berlin, Hinnerk van Bruinehsen
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     driver_periph
+ * @{
+ *
+ * @file
+ * @brief       Low-level UART driver implementation
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author      Hinnerk van Bruinehsen <h.v.bruinehsen@fu-berlin.de>
+ *
+ * @}
+ */
+
+#include "board.h"
+#include "cpu.h"
+#include "thread.h"
+#include "sched.h"
+
+#include "periph/uart.h"
+#include "periph_conf.h"
+
+/**
+ * @brief Allocate memory to store the callback functions.
+ */
+static uart_isr_ctx_t config[UART_NUMOF];
+
+static int init_base(uart_t uart, uint32_t baudrate);
+
+int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
+{
+    /* initialize basic functionality */
+    int res = init_base(uart, baudrate);
+
+    if (res != 0) {
+        return res;
+    }
+
+    /* register callbacks */
+    config[uart].rx_cb = rx_cb;
+    config[uart].arg = arg;
+
+    /* configure interrupts and enable RX interrupt */
+    switch (uart) {
+#if UART_0_EN
+
+        case UART_0:
+            UART0_RX_IRQ_EN;
+            break;
+#endif /* UART_0_EN */
+#if UART_1_EN
```

Comment:
There's only one UART on this CPU, so the code referring to uart interfaces > 0 should be removed.